cmd/create/idp/htpasswd.go

```diff
@@ -0,0 +1,179 @@
+		idpBuilder := cmv1.NewIdentityProvider().
+			Type("HTPasswdIdentityProvider").
+			Name(idpName).
+			Htpasswd(
+				cmv1.NewHTPasswdIdentityProvider().Users(
+					cmv1.NewHTPasswdUserList().Items(CreateHTPasswdUser(username, password)),
+				),
+			)
+		htpasswdIDP = doCreateIDP(idpName, *idpBuilder, cluster, clusterKey, ocmClient)
+	}
+
+	for shouldAddAnotherUser() {
+		username, password = getUserDetails(cmd, username, err, password)
+		err = ocmClient.AddHTPasswdUser(username, password, cluster.ID(), htpasswdIDP.ID())
+		if err != nil {
+			reporter.Errorf(
+				"Failed to add a user to the HTPasswd IDP of cluster '%s': %v", clusterKey, err)
+			os.Exit(1)
+		}
+		reporter.Infof("User '%s' added", username)
+	}
+}
+
+func getUserDetails(cmd *cobra.Command, username string, err error, password string) (string, string) {
+	username, err = interactive.GetString(interactive.Input{
+		Question: "Username",
+		Help:     cmd.Flags().Lookup("username").Usage,
+		Default:  "",
+		Required: true,
+		Validators: []interactive.Validator{
+			usernameValidator,
+		},
+	})
+	if err != nil {
+		exitHTPasswdCreate("Expected a valid username: %s", err)
+	}
+	password, err = interactive.GetPassword(interactive.Input{
+		Question: "Password",
+		Help:     cmd.Flags().Lookup("password").Usage,
+		Default:  "",
+		Required: true,
+	})
+	if err != nil {
+		exitHTPasswdCreate("Expected a valid password: %s", err)
+	}
+	return username, password
+}
+
+func shouldAddAnotherUser() bool {
+	addAnother, err := interactive.GetBool(interactive.Input{
+		Question: "Add another user",
+		Help:     "HTPasswd: Add more users to the IDP, to log into the cluster with.\n",
+		Default:  false,
+	})
+	if err != nil {
+		exitHTPasswdCreate("Expected a valid reply: %s", err)
+	}
+	return addAnother
+}
+
+func CreateHTPasswdUser(username, password string) *cmv1.HTPasswdUserBuilder {
+	builder := cmv1.NewHTPasswdUser()
+	if username != "" {
+		builder = builder.Username(username)
+	}
+	if password != "" {
+		builder = builder.Password(password)
+	}
+	return builder
+}
+
+func exitHTPasswdCreate(format string, err error) {
+	reporter.Errorf("Failed to create IDP for cluster '%s': %v",
+		clusterKey,
+		fmt.Errorf(format, err))
+	os.Exit(1)
+}
+
+func usernameValidator(val interface{}) error {
+	if username, ok := val.(string); ok {
+		if username == ClusterAdminUsername {
+			return fmt.Errorf("username '%s' is not allowed", username)
+		}
+		return nil
+	}
+	return fmt.Errorf("can only validate strings, got %v", val)
+}
+
+func HasClusterAdmin(htpasswdIDP *cmv1.IdentityProvider) bool {
+	hasAdmin := false
+	if htpasswdIDP != nil {
+		htpasswdIDP.Htpasswd().Users().Each(func(user *cmv1.HTPasswdUser) bool {
+			if user.Username() == ClusterAdminUsername {
+				hasAdmin = true
+			}
+			return true
+		})
+	}
+	return hasAdmin
+}
+
+func FindExistingHTPasswdIDP(cluster *cmv1.Cluster, ocmClient *ocm.Client) (htpasswdIDP *cmv1.IdentityProvider) {
```

Comment:
nitpick: I wonder if this function should be part of `pkg/ocm/idps.go`

Reply:
In the `ocm` package we wouldn't have access to the `idp` package's reporter. I could pass the reporter in as a parameter but this seems less correct to me.

Reply:
Makes sense, thanks